src/docs/arch-storage-extension.dox

```diff
@@ -0,0 +1,104 @@
+storage source reference is taken and decremented with the termination. When the reference count
+goes zero, the storage source is de-registered with WiredTiger and any allocated resources are
+freed.
+
+- Another essential functionality is that the storage source provides is to create a
+\c WT_FILE_SYSTEM implementation for a given bucket. The store requires a bucket and an access token
+to create the filesystem.
+
+@section extension_connection Connection class
+
+All extensions has a connection class which represents an active connection to a bucket. It
+encapsulates and implements all the operations the filesystem requires. The instantiation of the 
+class object authenticates with the the cloud provider and validates the existence of and access to
+the configured bucket. The connection class exposes an API to list the bucket contents filtered by a
+directory and a prefix, check for an object's existence in the bucket, put an object to the cloud,
+and get the object from the cloud. Though not required for the file system's implementation, the
+class also provides the means to delete the objects to clean up artifacts from the internal unit
+testing.
+
+@section extension_file_system FILE_SYSTEM
+
+WiredTiger supports a database over a custom file system by giving a \c WT_FILE_SYSTEM interface for
+the end-user to implement. The store itself provides a means to configure as many file systems as
+needed, each linked to a bucket. These file systems, called the \c GCP_FILE_SYSTEM, AZURE_FILE_SYSTEM 
+and S3_FILE_SYSTEM, are a custom implementation provided by the storage source. The filesystems 
+contains an instance to the connection class, hence providing an active connection to the bucket 
+embedded in the file system instance.
+
+The cloud file system uses the functionality provided by the connection class to simulate a
+write-once-read-many file system. The filesystem can list the directory contents, check for a
+file's presence, and access the file by creating a file handle. The files that are not in the cache
+are first downloaded from the bucket and put in the cache before a filehandle is made to access
+them.
+
+The storage source uses the filesystem's put-object functionality to flush the files into the cloud.
+The Tier Manager inside WiredTiger directs the flush of the files. Only S3 storage source files
+pushed to the cloud are also copied to the local cache for easy access.
+
+@section extension_file_handle FILE_HANDLE
+
+Each storage source also implements the \c WT_FILE_HANDLE interface, to access the files
+on the file system. Since the object store is archival, all the write access methods are disabled.
+
+@section extension_log_stat Logging and Statistics
+All of the cloud storage sources have their own logging system implemented. The storage sources
+provides a logger implemntation that redirects the generated logs to the WiredTiger logging system.
```

Comment:
Grammar: "The storage sources provides a logger" -> "The storage sources provide a logger". I would also consider replacing the term "logging" here, we typically use "messaging" to avoid any confusion with the write-ahead log.

Reply:
This might be worth more of a conversation. I think it's a little bit worse now that we're freely interchanging logging and messaging - I'd prefer to settle on messaging since that's the term we use elsewhere, but this section talks about logging a lot.